lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -0,0 +1,1395 @@
+package drivers
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pborman/uuid"
+	"golang.org/x/sys/unix"
+
+	"github.com/lxc/lxd/lxd/migration"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/revert"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/ioprogress"
+	"github.com/lxc/lxd/shared/units"
+)
+
+// CreateVolume creates an empty volume and can optionally fill it by executing the supplied
+// filler function.
+func (d *zfs) CreateVolume(vol Volume, filler *VolumeFiller, op *operations.Operation) error {
+	// Revert handling
+	revert := revert.New()
+	defer revert.Fail()
+
+	if vol.contentType == ContentTypeFS {
+		// Create mountpoint.
+		err := vol.EnsureMountPath()
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { os.Remove(vol.MountPath()) })
+	}
+
+	// Look for previously deleted images.
+	if vol.volType == VolumeTypeImage && d.checkDataset(d.dataset(vol, true)) {
+		// Restore the image.
+		_, err := shared.RunCommand("/proc/self/exe", "forkzfs", "--", "rename", d.dataset(vol, true), d.dataset(vol, false))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// After this point we'll have a volume, so setup revert.
+	revert.Add(func() { d.DeleteVolume(vol, op) })
+
+	if vol.contentType == ContentTypeFS {
+		// Create the filesystem dataset.
+		err := d.createDataset(d.dataset(vol, false), fmt.Sprintf("mountpoint=%s", vol.MountPath()), "canmount=noauto")
+		if err != nil {
+			return err
+		}
+
+		// Apply the size limit.
+		size := vol.ExpandedConfig("size")
+		if size != "" {
+			err := d.SetVolumeQuota(vol, size, op)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		// Convert the size.
+		size := vol.ExpandedConfig("size")
+		if size == "" {
+			size = defaultBlockSize
+		}
+
+		sizeBytes, err := units.ParseByteSizeString(size)
+		if err != nil {
+			return err
+		}
+
+		loopPath := filepath.Join(shared.VarPath("disks"), fmt.Sprintf("%s.img", d.name))
+		if d.config["source"] == loopPath {
+			// Create the volume dataset with sync disabled (to avoid kernel lockups when using a disk based pool).
+			err = d.createVolume(d.dataset(vol, false), sizeBytes, "sync=disabled")
+			if err != nil {
+				return err
+			}
+		} else {
+			// Create the volume dataset.
+			err = d.createVolume(d.dataset(vol, false), sizeBytes)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// For VM images, create a filesystem volume too.
+	if d.checkVMBlock(vol) {
+		fsVol := NewVolume(d, d.name, vol.volType, ContentTypeFS, vol.name, vol.config, vol.poolConfig)
+		err := d.CreateVolume(fsVol, nil, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { d.DeleteVolume(fsVol, op) })
+	}
+
+	// Mount the dataset.
+	_, err := d.MountVolume(vol, op)
+	if err != nil {
+		return err
+	}
+
+	if vol.contentType == ContentTypeFS {
```

Comment:
@stgraber looks like this block is a duplicate of the one at https://github.com/lxc/lxd/pull/6688/files#diff-889526b02b65fe324db3c38b4a6fa371R32

Reply:
It's not. The first one creates the mountpoint and sets the permissions on the mountpoint.
The second one sets the permission on the mount itself.

Reply:
@stgraber ah ok, is that a ZFS specific thing? Wondering if need to do something similar for LVM.

Reply:
LVM should do the same, so should CEPH.